### DISCUSSION TITLE
Account profile not loading through getProfile, but reading with at://.../app.bsky.actor.profile/self works

### DISCUSSION THREAD

**Discussion by thardie** [bluesky-social/atproto#3068] [Clarifications and "Why?"]
I'm running a few PDSes, and on one of them, a recently migrated user's profile stopped working. When using the web interface, this (authenticated) request to https://skyfurs.social/xrpc/app.bsky.actor.getProfile?actor=did%3Aplc%3A7rx7iquchjfd7kvmsbni3cax fails with a JSON response message of "Profile not found". If I use goat to query the profile with `goat record get at://did:plc:7rx7iquchjfd7kvmsbni3cax/app.bsky.actor.profile/self`, it returns a valid JSON object. I've looked through a bunch of the sqlite databases on the PDS trying to manually run the queries by looking at things like https://github.com/bluesky-social/atproto/blob/main/packages/pds/src/read-after-write/viewer.ts#L110 and re-doing the queries with sqlite3 on the command line. That results in the same data that I can see with goat, so I don't understand why app.bsky.actor.getProfile returns "Profile not found".

Other profiles on the same PDS load fine.

**Comment by yamarten** [comment]
When I saw it, the signature verification failed. There may have been a problem importing the repository.

https://pdsls.pages.dev/at/did:plc:7rx7iquchjfd7kvmsbni3cax/app.bsky.actor.profile/self

Unfortunately, I don't know how to solve this problem.

**Reply by DavidBuchanan314** [reply]
The bad signature error could imply that the did:plc was not updated properly

**Reply by DavidBuchanan314** [reply]
looking at the plc logs, the key *was* updated, but maybe it doesn't correspond to the key the PDS holds? https://plc.directory/did:plc:7rx7iquchjfd7kvmsbni3cax/log/audit

**Reply by thardie** [reply]
Thanks for pointing out the invalid signature. I'll see if I can fix it.

**Reply by thardie** [reply]
Signature is now fixed, but the original problem still persists.

**Reply by yamarten** [reply]
To make `getProfile` available, you need to make the relay recognize the update. Only if the API is used by the owner of the repository, the processing within the PDS can be completed by read-after-write.

This fix may be delayed until the next time the repository is updated.

**Reply by thardie** [reply]
> To make `getProfile` available, you need to make the relay recognize the update. Only if the API is used by the owner of the repository, the processing within the PDS can be completed by read-after-write.
> 
> This fix may be delayed until the next time the repository is updated.

So, is there anything I can do, or just wait? This particular account is not very old, and I could just tell the owner to delete and start again, but I'd like to figure this out from a technical standpoint so I could deal with it should it happen again in the future.

**Reply by yamarten** [reply]
Hmm, if my guess is correct, it will be fixed the next time it is updated. In other words, post, follow, or update your profile. If you want to fix it without them, it may be helpful to disable your account and activate it or send a requestCrawl, but it is unlikely.

**Reply by thardie** [reply]
> Hmm, if my guess is correct, it will be fixed the next time it is updated. In other words, post, follow, or update your profile. If you want to fix it without them, it may be helpful to disable your account and activate it or send a requestCrawl, but it is unlikely.

That's a good suggestion, except the app doesn't let you edit a profile when it can't find one. The "Edit profile" button is not there, instead just a page that says "Profile not found". I've tried editing the app.bsky.actor.profile entry with goat, and the update shows when I query the record, but the AppView still says "profile not found". I've tried deactivating and reactivating and requesting a re-crawl, and still same deal.

**Reply by yamarten** [reply]
The `rev` of the repository is not yet updated, so there is a high probability that the relay has not noticed. I'm not familiar with goat, but I wonder if the update was really reflected.

https://skyfurs.social/xrpc/com.atproto.sync.listRepos

I'm sorry if you can't see it from appview even if you update it, but there's nothing more I can do.

**Reply by thardie** [reply]
> The `rev` of the repository is not yet updated, so there is a high probability that the relay has not noticed. I'm not familiar with goat, but I wonder if the update was really reflected.
> 
> https://skyfurs.social/xrpc/com.atproto.sync.listRepos
> 
> I'm sorry if you can't see it from appview even if you update it, but there's nothing more I can do.

I'd done the update even before you suggested.

I've gone ahead, updated the profile again via goat and made a test post. Doesn't resolve it immediately, but I'll give it some time and see what happens.

**Comment by bnewbold** [comment]
I haven't dug in to this specific case, but we had a bunch of problems with relays and firehose around last Thursday (plus or minus a day). There is a chance events got dropped, or that PDS connection got dropped and things were left in a weird state.

I would try sending out new events from all relevant accounts: post, like, whatever. try updating the profile record (eg, add/remove whitespace), which can trigger re-indexing. check logs to see if `requestCrawl` is working.

**Reply by thardie** [reply]
> I haven't dug in to this specific case, but we had a bunch of problems with relays and firehose around last Thursday (plus or minus a day). There is a chance events got dropped, or that PDS connection got dropped and things were left in a weird state.
> 
> I would try sending out new events from all relevant accounts: post, like, whatever. try updating the profile record (eg, add/remove whitespace), which can trigger re-indexing. check logs to see if `requestCrawl` is working.

I've updated the profile and made a post for the one affected account. Other accounts on the PDS are working fine. It didn't resolve it immediately, but I'll wait a bit.

Thanks for the suggestions.

**Comment by thardie** [comment]
I spent a bunch more time trying to track this down, and stepping through code in the PDS tracked it down to the PDS making this call after it gets the call to https://skyfurs.social/xrpc/app.bsky.actor.getProfile?actor=did%3Aplc%3A7rx7iquchjfd7kvmsbni3cax :

https://api.bsky.app/xrpc/app.bsky.actor.getProfile?actor=did%3Aplc%3A7rx7iquchjfd7kvmsbni3cax

Which returns the "Profile not found" error.

I don't understand why this call is made. Why does the PDS ask the api.bsky.app for this profile? Surely api.bsky.app will just turn around and ask the PDS? This profile is stored on this PDS. How will the central service ever know this profile is updated?

I've done the suggestions above about making posts from the account and editing the profile. Still no change.

**Reply by yamarten** [reply]
The main reason is to fill in the information that the PDS does not have. The appview count followers, add labels, and replace icons with lightweight ones.

And the input to the appview is relayed by the relay. Updating the repository and sending `requestCrawl` is to let the relay know that there is information to be fetched.

**Reply by thardie** [reply]
> The main reason is to fill in the information that the PDS does not have. The appview count followers, add labels, and replace icons with lightweight ones.
> 
> And the input to the appview is relayed by the relay. Updating the repository and sending `requestCrawl` is to let the relay know that there is information to be fetched.

Oh, that makes sense. So, after half a dozen requestCrawls, profile updates, PLC updates, posts, and still the profile is "not found" - What now? Other accounts on the same PDS work fine....

**Comment by thardie** [comment]
I'm giving up on this. I deleted the profile and re-created.

**Comment by thardie** [comment]
There is still some buried problem. The new account I created with the same name (scotty.skyfurs.social) can be logged into fine. The profile can be edited, and works fine, except searching for scotty.skyfurs.social through AppView doesn't find it even after the account has existed for more than 24 hours. The new DID is did:plc:44rleyh3iqodwenovsulg6lt

It baffles me that this doesn't work like a DNS lookup. The AT protocol clearly describes how to resolve a name back to a DID immediately and get the PDS it's hosted on. Why doesn't AppView do this?

**Reply by thardie** [reply]
> There is still some buried problem. The new account I created with the same name (scotty.skyfurs.social) can be logged into fine. The profile can be edited, and works fine, except searching for scotty.skyfurs.social through AppView doesn't find it even after the account has existed for more than 24 hours. The new DID is did:plc:44rleyh3iqodwenovsulg6lt
> 
> It baffles me that this doesn't work like a DNS lookup. The AT protocol clearly describes how to resolve a name back to a DID immediately and get the PDS it's hosted on. Why doesn't AppView do this?

After confirming email and setting a profile picture, the account is now searchable on main. So, I've made the account work, but was never able to get the original profile issue stated here fixed, and was not able to use the original DID, so the person lost all their followers.
